website/docs/core-concepts/stacks/yaml-functions/terraform.output.mdx

```diff
@@ -0,0 +1,205 @@
+produce the result and cache it, and all the consecutive calls will just use the cached data. This is useful when you use the
+`!terraform.output` function for the same component in a stack in multiple places in Atmos stack manifests.
+It will speed up the function execution and stack processing.
+
+For example:
+
+<File>
+```yaml
+components:
+  terraform:
+    test2:
+      vars:
+        tags:
+          test: !terraform.output test {{ .stack }} id
+          test2: !terraform.output test {{ .stack }} id
+          test3: !terraform.output test {{ .stack }} id
+```
+</File>
+
+In the example, the `test2` Atmos component uses the outputs (remote state) of the `test` Atmos component from the same stack.
+The YAML function `!terraform.output` is executed three times (once for each tag).
+
+After the first execution, Atmos caches the result in memory,
+and reuses it in the next two calls to the function. The caching makes the stack processing much faster.
+In a production environment where many components are used, the speedup can be significant.
+
+## Security and Access Control Considerations
+
+When using the `!terraform.output` function (and `terraform output` command in general),
+addressing security and access control is critical to protect sensitive information.
+
+ - **Restrict Terraform CLI Access**: Only authorized personnel should have access to the system where the Terraform CLI is installed.
+   This prevents unauthorized users from running terraform output to retrieve sensitive data.
+
+ - **Role-Based Access Control (RBAC)**: Implement RBAC policies to ensure only those with a legitimate need can interact with Terraform projects and access outputs.
+
+ - **Audit Logs**: Maintain detailed logs of Terraform commands executed, including `terraform output`, to track who accessed sensitive information.
+
+ - **Output Restrictions**: Where possible, limit sensitive data exposure by avoiding the use of `terraform output` for secrets.
+   Store these values securely in dedicated secret management solutions (e.g., AWS Secrets Manager, HashiCorp Vault).
+
+ - **Encrypt State Files**: Ensure that Terraform state files (e.g., `terraform.tfstate`) are encrypted both at rest and in transit.
+   Use backend providers like AWS S3 with server-side encryption or other managed storage with encryption capabilities.
+
+ - **Limit State File Access**: Restrict access to the state file storage backend using IAM policies or similar controls to only authorized users and processes.
+
+ - **Remote State Locking**: Enable remote state locking to prevent accidental overwrites and ensure consistent state handling.
+   This is supported by backends like AWS DynamoDB when using S3.
+
+ - **Secrets in State Files**: Avoid exposing sensitive data in the state file whenever possible.
+   Mark sensitive outputs with `sensitive = true` in Terraform configurations, which prevents them from appearing in plaintext in the output or logs.
+
```

Comment:
This just reads like AI generated recommendations, that are not really that helpful.

Helpful recommendations are:
- Using `!terraform.output` with secrets will leak secrets to standard output (stdout) in any commands that describe stacks or components
- Using the function with `describe affected`, means Atmos will need access to all remote states used. If you have a narrow permissions scope (e.g. for dev) and use `describe affected` with stacks that reference production, it will fail.
- Overuse of the function within a stack for multiple components can signficantly reduce performance
- Consider DR implications if using it across regions
- Consider cold-start implications, when the dependent component is not yet provisioned, and `terraform output` will fail

```suggestion
When using the `!terraform.output` function (and `terraform output` command in general),
addressing security and access control is critical to protect sensitive information.

 - **Restrict Terraform CLI Access**: Only authorized personnel should have access to the system where the Terraform CLI is installed.
   This prevents unauthorized users from running terraform output to retrieve sensitive data.

 - **Role-Based Access Control (RBAC)**: Implement RBAC policies to ensure only those with a legitimate need can interact with Terraform projects and access outputs.

 - **Audit Logs**: Maintain detailed logs of Terraform commands executed, including `terraform output`, to track who accessed sensitive information.

 - **Output Restrictions**: Where possible, limit sensitive data exposure by avoiding the use of `terraform output` for secrets.
   Store these values securely in dedicated secret management solutions (e.g., AWS Secrets Manager, HashiCorp Vault).

 - **Encrypt State Files**: Ensure that Terraform state files (e.g., `terraform.tfstate`) are encrypted both at rest and in transit.
   Use backend providers like AWS S3 with server-side encryption or other managed storage with encryption capabilities.

 - **Limit State File Access**: Restrict access to the state file storage backend using IAM policies or similar controls to only authorized users and processes.

 - **Remote State Locking**: Enable remote state locking to prevent accidental overwrites and ensure consistent state handling.
   This is supported by backends like AWS DynamoDB when using S3.

 - **Secrets in State Files**: Avoid exposing sensitive data in the state file whenever possible.
   Mark sensitive outputs with `sensitive = true` in Terraform configurations, which prevents them from appearing in plaintext in the output or logs.

```